### PR TITLE
Raise codec errors when casting to std::string

### PR DESCRIPTION
**Pull request by sloretz**
## Description

This allows [the codec's exception](https://docs.python.org/3/c-api/unicode.html#c.PyUnicode_AsUTF8String) to be raised instead of `RuntimeError` when casting from `py::str` to `std::string`. I ran into this when porting a CPython extention to pybind11 with an existing test that expected a `UnicodeEncodeError` when given a string with [UCS surrogate characters](https://en.wikipedia.org/wiki/Universal_Character_Set_characters#Surrogates).

## Suggested changelog entry:

```rst
Allow the codec's exception to be raised instead of :code:`RuntimeError` when casting from :code:`py::str` to :code:`std::string`.
```

### CODE REVIEW

File: tests/test_pytypes.py [hunk 1]
```diff
@@ -133,6 +133,14 @@ def __repr__(self):
     else:
         assert m.str_from_handle(malformed_utf8) == "b'\\x80'"
 
+    assert m.str_from_string_from_str("this is a str") == "this is a str"
+    ucs_surrogates_str = "\udcc3"
+    if env.PY2:
+        assert u"\udcc3" == m.str_from_string_from_str(ucs_surrogates_str)
```
Comment (EricCousineau-TRI) [inline tests/test_pytypes.py:139]: Howdy Shane, thanks!
Any chance you know what's going wrong in the Python 2.x code?
```
UnicodeEncodeError: 'ascii' codec can't encode character u'\udcc3' in position 38: ordinal not in range(128)
```
https://github.com/pybind/pybind11/pull/2903/checks?check_run_id=2115679652#step:11:98

I can take a deeper look if ya want.

**Review by Skylion007 — APPROVED** [review]

**Comment by rwgk** [comment]
Hi @Skylion007, consider this approved also by me, but ideally I'd run this through the Google-internal global testing (which includes testing a large number of open source packages) before merging. Is it OK with you to wait a few hours?

In the past we had a couple cases with a surprising number of hiccups caused by similar changes. I feel this one will be fine, but good to be sure by testing first.

**Comment by Skylion007** [comment]
@rwgk Sure, go ahead.

**Comment by rwgk** [comment]
> @rwgk Sure, go ahead.

The testing just finished: no issues!

I'll go ahead with merging this PR. Thanks everybody! And sorry @sloretz this took us so long.

**Review by rwgk — APPROVED** [review]

**Pull request merged**